### ISSUE TITLE
FS: em_iterations is diagnosed but unreachable (link_threshold half shipped in #2636; see status comment)

### ISSUE THREAD

**Issue by benzsevern** [benzsevern/goldenmatch#2637]
Measured in #2636 (harnesses at `scripts/fs_levers/`). The goal for FS autoconfig is not to set every lever, it is to have the **ability** to reach each one and **know when** to. Two levers fail that test in the same specific way: the engine emits a correct runtime diagnosis and nothing can respond to it.

## 1. `link_threshold` — inert by construction

`auto_configure_probabilistic_df` never sets it, so:

```
mk.link_threshold = None  ->  mk.cutoff = None
                          ->  _perturbable_matchkeys() == []
                          ->  ThresholdShift.apply() returns None
```

Verified INERT on **all four** locally-available datasets (synthetic, anchor_person_match, ncvr_synthetic, historical_50k), so this is structural, not data-dependent. On a zero-config FS run the healer/optimizer has **no working FS-specific lever at all** — `threshold_shift` is the only member of the closed `ConfigEdit` vocabulary that reaches anything FS-specific, and it no-ops. The other five ops are shared with the weighted path.

Two consequences:

- The healer cannot explore the single most consequential FS decision.
- `threshold_perturbations()` returns `[]`, so `perturbation_stability` is `None` forever on the FS path. That is *honest* (the #2483 fix made it return `None` rather than fake stability) but an entire zero-label confidence signal never fires.

Meanwhile the engine already knows, and says so:

> *"linked 89.2% of records into multi-member clusters using a FALLBACK link cutoff of 0.5000 -- a fixed default, not a value chosen from this data (no 'link_threshold' was set and EM produced no calibrated cutoff) ... Set 'link_threshold' on the matchkey to choose it explicitly. See #2483."*

The detector works. It is wired to the user, not to the healer.

### Why the obvious fix does not work

"Have autoconfig set `link_threshold` from the EM-calibrated cutoff" is **not buildable as stated**, for two independent reasons:

1. `calibrated_link_threshold` is produced by `_calibrate_link_threshold` behind `GOLDENMATCH_FS_CALIBRATE_THRESHOLD`, which is **off by default** — hence "EM produced no calibrated cutoff" in the warning above.
2. Calibration happens during EM training, which runs *after* `auto_configure_probabilistic_df` has already returned a config. Autoconfig has no `EMResult` to read.

### The fix that does work

The **controller** is the right place, because it already trains. `_run_pipeline_sample` goes through `resolve_thresholds(mk, em_result)` (`pipeline.py:90`), and `link_threshold_source(mk, em_result)` (`probabilistic.py:3243`) already classifies the result as `configured` / `calibrated` / `fallback`. So at commit time the controller holds both the resolved number and its provenance.

Proposal: when the committed config's source is `calibrated` or `fallback`, stamp the resolved value onto `mk.link_threshold`. Effects:

- Default behavior is **unchanged** — the same number is used, it is just written down.
- `mk.cutoff` becomes non-`None`, so `ThresholdShift` goes live and the healer gains a real FS lever.
- `perturbation_stability` starts being measured on the FS path.
- The config autoconfig hands the user stops being silent about its most important decision, which is the actual user-facing complaint in #2483.

Worth care: this touches the controller's commit path, so it changes what every zero-config FS config *contains*. It deserves its own PR with tests, not a tack-on.

## 2. `em_iterations` / `convergence_threshold` — same shape

`EM did not converge after 20 iterations (delta=0.001560)` fired on exactly **one cell of six** in the probe: the `historical_50k levels=5` run that collapsed to F1 0.4803. Zero false positives in that sample. Nothing can raise the iteration budget in response, and neither field is set by autoconfig or reachable by any edit.

I had previously argued these were *correctly* unreachable, on the grounds that solver hyperparameters are noise-chasing. The measurement says otherwise: there is a clean, rare, directional distress signal attached to them, which is exactly the trigger such a lever would need.

## What is NOT proposed

**No `LevelsEdit`.** `levels` has the largest measured effect of anything here (F1 0.48–0.98) but its best value is dataset-dependent — `levels=5` is +0.0166 on ncvr_synthetic and −0.3406 on historical_50k — and the signal that looks like its trigger, the non-monotonicity warning, **points the wrong way**: it fires on 5 of 6 cells including both baselines, and ncvr's *best* cell has the *most* non-monotonic fields.

Giving the healer a lever of that magnitude with no way to aim it puts a −0.34 F1 move inside its reach. `levels=3` is the max-min choice across both datasets, so the current hardcoded default is defensible. This needs a real trigger before it needs an edit.

## Limits on the evidence

Two informative datasets, six cells. `dblp_acm`, `febrl3` and `ncvr_real` are absent locally and the anchors saturate at F1 1.0. Enough to identify the gaps, not enough to tune any threshold. The `link_threshold` finding is the exception: it is structural and reproduces on 4/4 datasets.

## Reproduce

```bash
uv sync --package goldenmatch --extra polars
python scripts/fs_levers/lever_sweep.py  --datasets ncvr_synthetic historical_50k --row-cap 20000 --out sweep.json
python scripts/fs_levers/trigger_probe.py --datasets ncvr_synthetic historical_50k --row-cap 20000 --out triggers.json
```

**Comment by benzsevern** [comment]
## Status update: the proposed fix shipped, but narrower than proposed, and it caused a real defect at scale

This issue's proposal was: *"when the committed config's source is
`calibrated` or `fallback`, stamp the resolved value onto `mk.link_threshold`."*
That shipped as `_stamp_resolved_link_thresholds`, and then caused a defect
worth being explicit about here.

**What went wrong.** Stamping a `fallback` value pins `mk.link_threshold`, and
once it's set the matchkey source becomes `configured` -- which skips the
full-data threshold refit entirely (`resolve_thresholds` short-circuits on an
explicit value). `fallback` means nothing chose that number; it's a fixed
default resolved against a ~6K sample. Pinning it costs real precision:
measured at person@100K, precision was **0.9308** with the fallback stamped
vs **0.9992** on the probabilistic lane that left it alone and let the
full-data refit run. Full writeup:
`docs/superpowers/notes/2026-08-18-zeroconfig-recall-incident.md` (defect 4).

**The fix I shipped narrows the stamp to `calibrated`-only** --
`fallback` is left `None` so the refit still runs. That resolves the defect,
but it means this issue's original problem is **only half-fixed**:

- Runs where EM produces a `calibrated` cutoff: fixed as proposed. The healer
  gets a working `link_threshold` lever, `_perturbable_matchkeys()` is
  non-empty, `perturbation_stability` can fire.
- Runs where the source is `fallback` (the common case, since
  `GOLDENMATCH_FS_CALIBRATE_THRESHOLD` is off by default per this issue's own
  text): **exactly as inert as before this issue was opened.**
  `mk.link_threshold` stays `None`, `_perturbable_matchkeys()` stays empty,
  `threshold_perturbations()` stays `[]`.

So this issue shouldn't read as closed by the merge that references it. The
calibrated half is resolved; the fallback half -- which by the issue's own
math is the more common one -- needs its own answer. Recording a real cutoff
without pinning it (i.e. keep it explicit for the user/telemetry but let
`resolve_thresholds` still treat it as needing a refit) would need a third
state beyond the current `configured` / `calibrated` / `fallback`, since
`configured` is what currently means "don't refit" and that's precisely the
behavior a stamped fallback must NOT trigger. I don't have a design for that
split yet -- flagging it rather than proposing one under the same pressure
that produced the original bug.

**Comment by benzsevern** [comment]
Re-verified against `main` @ `ffc8ca917` (2026-08-21). The two halves have diverged and the title no longer describes what is left.

## Half 1 -- `link_threshold`: mostly closed, and the residual is deliberate

`_stamp_resolved_link_thresholds` is on main and its docstring cites this issue. Verified in code:

- `calibrated` **is** stamped, so `mk.cutoff` is set, `_perturbable_matchkeys()` is non-empty, and `ThresholdShift` reaches the FS path.
- `configured` is preserved (the user decided).
- `fallback` is **deliberately not** stamped, per the comment above: pinning it makes the source `configured`, which short-circuits `resolve_thresholds` and skips the full-data refit -- measured at pairwise precision 0.9308 vs 0.9992 on person@100K.

So "inert on 4 of 4 datasets" is no longer true, and the remaining inertness is a measured design constraint rather than a gap. What is genuinely left here is narrow: **on a zero-config FS run where EM produces no calibrated cutoff, the healer still has no FS-specific lever.** That is a real hole, but it is the fallback case only, and the fix cannot be "stamp it anyway".

## Half 2 -- `em_iterations` / `convergence_threshold`: untouched

Verified on main: **zero** references in `core/autoconfig*.py` or `core/config_edits.py`. The `ConfigEdit` vocabulary is still `BlockingStrategyEdit` / `BlockingKeyEdit` (+ the shared ops). Neither field is set by autoconfig or reachable by any edit, exactly as filed. `EM did not converge after 20 iterations` still has nothing that can respond to it.

## Suggested re-scope

This should become **"`em_iterations` is diagnosed but unreachable"**, carrying half 2 plus the one-line residual from half 1. The original framing -- both levers structurally inert -- describes a `main` that no longer exists, and anyone picking it up would re-derive the `link_threshold` work that already shipped.

Not closing it: half 2 reproduces exactly as written.

**Comment by benzsevern** [comment]
## Measured: the `em_iterations` case is refuted, on the cell that motivated it

This issue's argument for an `em_iterations` lever rests on a trigger — *"a clean, rare, directional distress signal ... exactly the trigger such a lever would need"* — and on one untested assumption behind it: that **raising the budget moves the cell**.

Harness committed at `scripts/fs_levers/em_budget_probe.py`, output under `docs/measurements/`. `historical_50k`, `levels=5`, `--row-cap 20000`, only `em_iterations` varied:

```
em_iterations   converged   F1        precision   recall
            1   NO          0.5109    0.3555      0.9075
           20   yes         0.4744    0.3282      0.8557
          100   yes         0.4744    0.3282      0.8557
```

**1. Raising the budget changes nothing.** 20 → 100 is identical to four decimal places on every metric.

**2. The trigger does not fire on this cell at the default budget.** `converged=True` at 20. The distress signal this issue's case is built on does not reproduce on current `main`. It was measured 2026-08-17; something between then and now changed it. Recording that rather than explaining it.

**3. Less EM is better.** `em_iterations=1` — maximally unconverged, precisely the state the proposed lever exists to escape — scores the highest F1 of the three.

So the direction is **inverted**. This cell is not a solver that ran out of road; it is a model misspecified at `levels=5`, and EM converges confidently onto a bad solution. An "increase the budget when EM does not converge" lever would move F1 the **wrong way** on the only cell that motivated it.

That is the same shape as the argument this issue itself makes for not adding a `LevelsEdit`: *"the signal that looks like its trigger points the wrong way."* It applies here too — which means the position stated and then abandoned in this issue ("solver hyperparameters are noise-chasing") was right, and the #2636 measurement that overturned it was reading a coincidence.

The failure this cell actually has is precision 0.33 at recall 0.86 — heavy over-merging. That is a threshold and level-specification problem, not an EM budget one.

### On trusting the instrument

The harness carries a **known-positive**: `em_iterations=1` cannot converge, so if that run reports convergence the log handler is broken and every "converged" reading is worthless. It fires (`delta=0.445971`). Without it, finding 2 would have been an unfalsifiable null — "no warning" and "no listener" look identical.

### What remains

**`link_threshold`'s `fallback` half is untouched by this.** The status comment above is still accurate: `calibrated` runs are fixed, `fallback` runs are as inert as before, and closing that needs a way to record a resolved cutoff without it counting as `configured` — because `configured` is what currently means "skip the full-data refit", and that is exactly the behaviour that cost precision 0.9308 vs 0.9992 at person@100K.

I traced that to something narrower than a new threshold state: `mk.link_threshold` currently carries two meanings at once — the *value*, and the assertion "a human chose this, do not refit". The healer only needs `mk.cutoff is not None`; `resolve_thresholds` only needs to know whether a human decided. Separating **provenance** from **value** is the smaller change. It does touch the config schema, which is a public surface with TypeScript parity and generated config docs behind it, so it is a deliberate piece of work rather than a tack-on.

Not filing that separately — it belongs to this issue, and this issue stays open for it.
